File: Roadmap/03 - ESTRUCTURAS DE DATOS/javascript/agusrosero.js
/*
 * EJERCICIO:
 * - Muestra ejemplos de creación de todas las estructuras soportadas por defecto en tu lenguaje.
 * - Utiliza operaciones de inserción, borrado, actualización y ordenación.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea una agenda de contactos por terminal.
 * - Debes implementar funcionalidades de búsqueda, inserción, actualización y eliminación de contactos.
 * - Cada contacto debe tener un nombre y un número de teléfono.
 * - El programa solicita en primer lugar cuál es la operación que se quiere realizar, y a continuación
 *   los datos necesarios para llevarla a cabo.
 * - El programa no puede dejar introducir números de teléfono no númericos y con más de 11 dígitos.
 *   (o el número de dígitos que quieras)
 * - También se debe proponer una operación de finalización del programa.
 */

// EJERCICIO:
// Arrays
let miArray = [1, 2, 3, 4, 5];
miArray.push(10);
miArray.pop();
miArray[0] = 23;
miArray.sort();
console.log(miArray);

// Sets
const miSet = new Set([1, 2, 3]);
miSet.add(4);
miSet.delete();
console.log(miSet);

// Objects
const miObjeto = {
  x: 10,
  y: 15,
  z: 20,
};
miObjeto.punto = 2;
delete miObjeto.z;
miObjeto["y"] = 45;
console.log(miObjeto);

// Maps
const miMap = new Map([
  ["a", 1],
  ["b", 2],
]);
miMap.set("c", 3);
miMap.delete();
miMap.set("a", "a");
console.log(miMap);

// DIFICULTAD EXTRA:
const agendaContactos = () => {
  let agenda = new Map();

  let opcion = 1;
  let nombre = "Hernan";
  let numero = 3413646464;
  let nombre2 = "Agustin";
  let numero2 = 3413646566;

  console.log(`
    1. Buscar
    2. Insertar
    3. Actualizar
    4. Eliminar
    5. Salir
  `);

  if (opcion >= 1 && opcion <= 5) {
    switch (opcion) {
      case 1:
        if (nombre !== "" && !isNaN(numero) && numero.toString().length <= 11) {
          agenda.set(nombre, numero);
          console.log(agenda);
        } else {
          console.log("Error, vuelva a insertar datos validos");
        }
        break;
      case 2:
        if (nombre !== "" && !isNaN(numero) && numero.toString().length <= 11) {
          agenda.set(nombre, numero);
        } else if (agenda.has(nombre)) {
          console.log(`El numero de ${nombre} es: ${agenda.get(nombre)}`);
        } else {
          console.log("El contacto no esta registrado");
        }
        break;
      case 3:
        agenda.set(nombre2, numero2);
        console.log(`El numero de ${nombre2} es: ${agenda.get(nombre2)}`);
        break;
      case 4:
        agenda.delete(nombre);
        console.log(agenda);
        break;
      case 5:
        console.log("Saliendo de la agenda..");
        break;
      default:
        break;
    }
  } else {
    console.log("Error de opcion, ingrese un numero entre el 1 y el 5");
  }
};

agendaContactos();

File: Roadmap/03 - ESTRUCTURAS DE DATOS/java/agusrosero.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Scanner;
import java.util.Set;

public class agusrosero {
    public static void main(String[] args) {
        // Arrays
        String[] array = new String[3];

        array[0] = "Hola";
        array[1] = "Mundo";
        array[2] = "Java";

        // Lista
        List<String> list = new ArrayList<>();
        list.add("Hola");
        list.add("Mundo");
        list.add("Java");

        list.clear();

        // Set
        Set<String> set = new HashSet<>();

        set.add("Hola");
        set.add("Mundo");
        set.add("Java");

        set.clear();

        // Map
        Map<String, String> hashMap = new HashMap<>();

        hashMap.put("Hola", "Mundo");
        hashMap.put("Java", "Kotlin");

        hashMap.clear();

        /*
         * * DIFICULTAD EXTRA (opcional):
         * Crea una agenda de contactos por terminal.
         * - Debes implementar funcionalidades de búsqueda, inserción, actualización y
         * eliminación de contactos.
         * - Cada contacto debe tener un nombre y un número de teléfono.
         * - El programa solicita en primer lugar cuál es la operación que se quiere
         * realizar, y a continuación
         * los datos necesarios para llevarla a cabo.
         * - El programa no puede dejar introducir números de teléfono no númericos y
         * con más de 11 dígitos.
         * (o el número de dígitos que quieras)
         * - También se debe proponer una operación de finalización del programa.
         */
        System.out.println("*** Agenda de contactos ***");
        System.out.println("1. Buscar contacto");
        System.out.println("2. Insertar contacto");
        System.out.println("3. Actualizar contacto");
        System.out.println("4. Eliminar contacto");
        System.out.println("5. Salir");

        Scanner scanner = new Scanner(System.in);
        int option = 0;
        List<String> contacts = new ArrayList<>();

        while (true) {
            System.out.print("Seleccione una opcion: ");
            option = scanner.nextInt();
            switch (option) {
                case 1:
                    System.out.print("Ingrese un nombre: ");
                    String search = scanner.next();
                    for (String contact : contacts) {
                        if (contact.contains(search)) {
                            System.out.println(contact);
                        } else {
                            System.out.println("Contacto no encontrado");
                        }
                    }
                    break;
                case 2:
                    System.out.print("Ingrese un nombre: ");
                    String name = scanner.next();
                    System.out.print("Ingrese un número de teléfono: ");
                    String phoneNumber = scanner.next();
                    contacts.add(name + " - " + phoneNumber);
                    break;
                case 3:
                    System.out.print("Ingrese el nombre del contacto a actualizar: ");
                    String oldName = scanner.next();
                    System.out.print("Ingrese el nuevo nombre: ");
                    String newName = scanner.next();
                    System.out.print("Ingrese el nuevo número de teléfono: ");
                    String newPhoneNumber = scanner.next();
                    for (String contact : contacts) {
                        if (contact.contains(oldName)) {
                            contacts.remove(contact);
                            contacts.add(newName + " - " + newPhoneNumber);
                        }
                    }
                    break;
                case 4:
                    System.out.print("Ingrese el contacto a eliminar: ");
                    String contact = scanner.next();
                    contacts.remove(contact);
                    break;
                case 5:
                    System.out.println("Saliendo...");
                    scanner.close();
                    System.exit(0);
                    break;
                default:
                    System.out.println("Opción no válida");
            }
        }
    }
}
